L112-node-server-interceptors.md

```diff
@@ -0,0 +1,244 @@
+   */
+  onReceiveMetadata?: (metadata: Metadata, next: (metadata: Metadata) => void): void;
+  /**
+   * An inbound interception method called when each message is received.
+   */
+  onReceiveMessage?: (message: any, next: (message: any) => void): void;
+  /**
+   * An inbound interception method called when the inbound message stream ends.
+   */
+  onReceiveHalfClose?: (next: () => void): void;
+  /**
+   * An inbound interception method called when the request ends for any reason, including when cancelled by the remote end.
+   */
+  onCancel?: (next: () => void): void;
+}
+```
+
+A `ServerListenerBuilder` will be provided to easily construct a listener:
+
+```javascript
+const listener = (new ServerListenerBuilder())
+      .withOnReceiveMetadata(function(metadata, next) {
+          logger.log(metadata);
+          next(metadata);
+      })
+      .withOnReceiveMessage(function(message, next) {
+          logger.log(message);
+          next(message);
+      })
+      .build();
+```
+
+#### Correspondence with client interceptor APIs
+
+The server interceptor APIs are roughly mirrors of the client interceptor APIs, in the sense that send operations on the client correspond to receive operations on the server, and vice versa. In particular, the methods are matched this way as follows:
+
+| Client API | Server API | Notes |
+| ---------- | ---------- | ----- |
+| `Requester.start` | `ServerListener.onReceiveMetadata` | |
+| `Requester.sendMessage` | `ServerListener.onReceiveMessage` | |
+| `Requester.halfClose` | `ServerListener.onReceiveHalfClose` | |
+| `Requester.cancel` | `ServerListener.onCancel` | This server API also handles calls ending in ways other than client cancellation. |
+| | `Responder.start` | This is a utility function for setting up interceptors that does not correspond to any client action. |
+| `Listener.onReceiveMetadata` | `Responder.sendMetadata` | |
+| `Listener.onReceiveMessage` | `Responder.sendMessage` | |
+| `Listener.onReceiveStatus` | `Responder.sendStatus` | |
+
+#### Internal Errors and Exceptions
+
+Exceptions are not handled by the interceptor framework, we expect
+interceptor authors to either handle the exceptions their interceptors
+can throw or assume the Node process will crash on exceptions. This is in
+accordance with the [Joyent guidance](https://www.joyent.com/node-js/production/design/errors)
+to only catch exceptions which are known to be operational errors.
+
+In the case of an error during the execution of an interceptor,
+interceptor authors have the following options:
+1) Short-circuiting the call and returning a gRPC error to the client's consumer.
+2) Throwing an exception which the client's consumer will have to catch
+or crash on.
+3) Doing nothing and continuing with the call.
+
+### Use
+
+The `Server` constructor's `options` parameter will accept a new option `interceptors`, which is an array of interceptor functions.
+
+
+#### Interceptor nesting
```

Comment:
Nit: chaining may be more commonly used for this term.

One of the reasons I asked about the client interceptor design is to find out if I need to review the interceptor chaining or nesting flow.

E.g.  if interceptorA onReceiveMetadata() decides to respond a status immediately (e.g. an auth failure) will  interceptorB/C onReceiveMetadata()  be called by the framework or the call stack will jump to the outbound flow (for interceptorA).

Reply:
I used "nesting" because it's the term that the [corresponding section of the client interceptors design](https://github.com/grpc/proposal/blob/master/L5-node-client-interceptors.md#interceptor-nesting) used.

The `next` callback in each method allows the interceptor author to control that flow. In your particular auth failure example, the auth interceptor's `onReceiveMetadata` method can explicitly send a non-OK status and not call the `next` callback. As a result, the entire rest of the inbound/outbound flow will be skipped, and the handler will never be invoked. This is broadly the same as in the client interceptors design.